### PR TITLE
 Issue #529: Deprecate launch.groovy and merge functionality into diff.groovy

### PR DESCRIPTION
**Pull request by nmancus1**
Issue #529: Deprecate launch.groovy and merge functionality into diff.groovy

Notes:

-  We should replace all usages of `launch.groovy` with `diff.groovy` in single mode, using 'patchBranch' and 'patchConfig' . This isn't a big deal, we are just saving an extra checkstyle run using `master` branch.
- Any usages of `launch` that are being replaced with `diff` that DO NOT use `-i` need to specify `--allowExcludes` now.
- All existing `diff.groovy` usage needs to be reviewed; `no-exception` tests can be done in single mode.

Related PRs that must be merged after this for CI in other projects:
https://github.com/checkstyle/checkstyle/pull/9278
https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/828

A good example of `no-exception` testing and replacing `launch.groovy` for many cases is:
https://github.com/checkstyle/checkstyle/blob/492dd22e4730b7902e0af91819da840c2dac7b23/.ci/no-exception-test.sh#L68

### CODE REVIEW

**Comment by rnveach** [comment]
> Any existing diff usages do not need to change.

You need to review that one I pointed out not too long ago that was converted from launch to diff prematurely before this PR.

**Comment by nmancus1** [comment]
> You need to review that one I pointed out not too long ago that was converted from launch to diff prematurely before this PR.

Yes, you are right, that one can be changed to single mode too, as all no-exceptions testing can.

**Comment by nmancus1** [comment]
At this point CI is green for all PR's related to this change in Checkstyle and Sevntu:
https://github.com/checkstyle/checkstyle/pull/9278
https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/828

Still need to change over CI scripts in both to do single mode, instead of diff mode to save CI time.

**Comment by nmancus1** [comment]
#### AppVeyor:
___
DESC=`checkstyle-tester (diff.groovy) on guava`
maven command from master: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`
maven command from master: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`

maven command from PR: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`
maven command from PR: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`
___
DESC=`checkstyle-tester (diff.groovy with base and patch configs) on guava`
maven command from master: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`
maven command from master: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`

maven command from PR: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`
maven command from PR: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`


#### Travis:
___
CMD=`./.ci/travis.sh checkstyle-tester-launch-groovy`
maven command from master: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.failsOnError=false`
maven command from PR: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`
___
CMD="./.ci/travis.sh checkstyle-tester-diff-groovy-patch"
maven command from master: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`
maven command from master: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`
maven command from PR: ` mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`
maven command from PR: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`
___
CMD=`./.ci/travis.sh checkstyle-tester-diff-groovy-patch-only`
maven command from master: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=fals`

maven command from PR: `mvn -e --batch-mode site -Dcheckstyle.config.location=my_check.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.41-SNAPSHOT -Dcheckstyle.failsOnError=false`

File: checkstyle-tester/README.md
Comment (nmancus1) [inline checkstyle-tester/README.md:76]: Now we can allow excludes in diff.groovy execution

**Comment by romani** [comment]
Does it make sense to remove launch.groovy ?

**Comment by nmancus1** [comment]
> Does it make sense to remove launch.groovy ?

Probably. I can do it as a separate commit (with corresponding documentation change) in case we need to revert it.

@rnveach what do you think?

File: .ci/travis.sh
Comment (rnveach) [inline .ci/travis.sh:74]: The way i understand this, you run regression using Contribution Master and Contribution PR and check if the 2 separate runs produce the same or different files. I assume difference will fail the CI (and are we sure of that?).

If the above is correct, is this planned to be merged in? What happens if we make diff-report-util change the layout. This will then always fail CI even though it is an expected change.

File: .ci/travis.sh
Comment (rnveach) [inline .ci/travis.sh:85]: Whats the difference compared to https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/checks-nonjavadoc-error.xml ?

File: checkstyle-tester/README.md
Comment (rnveach) [inline checkstyle-tester/README.md:70]: Line starting with just "`" seems odd.

**Comment by nmancus1** [comment]
> I think it would be safer to have it in it's own PR to be sure it is ready to be removed.

No complaints here, I can just submit a PR for launch removal once this is merged.

File: checkstyle-tester/README.md
Comment (rnveach) [inline checkstyle-tester/README.md:75]: Instead of saying some unknown `projects*.properties` file which doesn't even account for how it fits into the run the user is trying to make, I would say sopmething along that this connects to the same property file passed into `listOfProjects (l)`.

Connect the dots for the user.

File: checkstyle-tester/README.md
Comment (rnveach) [inline checkstyle-tester/README.md:78]: Maybe I lost track. Why do we need to pass in a checkstyle version to override what is being used automatically? Point to example if there is one.

File: checkstyle-tester/README.md
Comment (rnveach) [inline checkstyle-tester/README.md:81]: Same here. We are not saying this supports sevntu regression, right?

**Review by rnveach — COMMENTED** [review]
First pass questions. Haven't looked at code in depth.

File: checkstyle-tester/diff.groovy
Comment (rnveach) [inline checkstyle-tester/diff.groovy:53]: Documentation still says this is required. Why did we change this to optional?

File: checkstyle-tester/diff.groovy
Comment (rnveach) [inline checkstyle-tester/diff.groovy:57]: Same.

**Review by rnveach — COMMENTED** [review]
2 more connected to documentation as I go through code.

File: .ci/travis.sh
Comment (nmancus1) [reply .ci/travis.sh:74]: > I assume difference will fail the CI (and are we sure of that?).

Yes. https://travis-ci.org/github/checkstyle/contribution/jobs/763804426

> If the above is correct, is this planned to be merged in? What happens if we make diff-report-util change the layout. This will then always fail CI even though it is an expected change.

I would guess at that point we would need to use some sort of more complicated awk or perl command to verify that violations match up; right now, this CI job confirms that both the violations and format of the report are consistent.

File: .ci/travis.sh
Comment (nmancus1) [reply .ci/travis.sh:85]: Right, I basically copied that config and made it regression compatible.

File: checkstyle-tester/README.md
Comment (nmancus1) [reply checkstyle-tester/README.md:78]: https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/828/files#r599509434

File: checkstyle-tester/README.md
Comment (nmancus1) [reply checkstyle-tester/README.md:81]: https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/828/files#r599509434

File: checkstyle-tester/diff.groovy
Comment (nmancus1) [reply checkstyle-tester/diff.groovy:53]: Fixed, see 
https://gist.github.com/nmancus1/bae2ef94de5d0457c8c59631d5192b88 for new behavior (added checking for both of the mentioned args in diff mode).

This is optional now because of sevntu usage.

File: checkstyle-tester/diff.groovy
Comment (nmancus1) [reply checkstyle-tester/diff.groovy:57]: Same as above.

File: checkstyle-tester/README.md
Comment (nmancus1) [reply checkstyle-tester/README.md]: Connected.

File: checkstyle-tester/README.md
Comment (nmancus1) [reply checkstyle-tester/README.md]: Fixed.

**Comment by nmancus1** [comment]
AppVeyor failure: `"Unable to connect to the remote server"`.

File: checkstyle-tester/diff.groovy
Comment (nmancus1) [reply checkstyle-tester/diff.groovy]: These review items might have been on a cached version of diff, I removed these with last push.

File: checkstyle-tester/diff.groovy
Comment (nmancus1) [reply checkstyle-tester/diff.groovy]: Same as above.

File: checkstyle-tester/README.md
Comment (rnveach) [reply checkstyle-tester/README.md:79]: ok, this is just for no exception testing in sevntu.

These 2 properties are pretty dangerous to use. If the wrong version is set, maven won't fail normally. It will try to find this version installed in the repo or pulled from the server. If it finds a version to use, it will use that and not be noticeable to the user.

Take our own regression, someone active in the repo will always be running local regression. So it is possible they have previous versions installed. If the use this and set an old version, the report will be generated however it will not be hitting their new code at all and only hitting old. This will give a false sense of testing, especially if no changes are to be expected.

I would add some verbage to this to make it clear it is not to be used in normal circumstances. It is for internal use only.

Ideally, I think these should be removed but I don't see an alternative to get no exception testing to work in sevntu. If support is added for sevntu, these properties should then be removed.

File: .ci/travis.sh
Comment (rnveach) [reply .ci/travis.sh:74]: I wouldn't do anything too complicated. Should we not merge it or is there some alternative for this to not cause issues easily?

File: checkstyle-tester/README.md
Comment (rnveach) [reply checkstyle-tester/README.md:82]: Same here.

File: checkstyle-tester/README.md
Comment (nmancus1) [reply checkstyle-tester/README.md:79]: I can add a check to make sure that these are only used in single mode, too.

File: checkstyle-tester/diff.groovy
Comment (rnveach) [reply checkstyle-tester/diff.groovy:53]: > This is optional now because of sevntu usage.

I understand but I don't fully agree. This is only because of no sevntu support. We would have to flop this is support is ever given as repo is always required.

Can we change sevntu usage to give sevntu repo and just not do anything with it? Maybe a different alternative.

File: .ci/travis.sh
Comment (nmancus1) [reply .ci/travis.sh:74]: I think we should keep these, I have pushed several times in this PR, and these jobs haven't been flaky. It is the best way that I can think of to easily check that the violations, reports, and report formats are all consistent.

I think it is ok to wait and see what the future brings regarding changes to `patch-diff-util` before doing anything different here.

File: checkstyle-tester/README.md
Comment (nmancus1) [reply checkstyle-tester/README.md:82]: > Same here.

What is this comment referring to?

File: checkstyle-tester/README.md
Comment (nmancus1) [reply checkstyle-tester/README.md:79]:
```bash

➜  checkstyle-tester git:(issue-529) ✗ groovy diff.groovy -l projects-to-test-on.properties -r ../../checkstyle -p my-issue --checkstyleVersion "8.41.1"

Picked up _JAVA_OPTIONS: -Dawt.useSystemAAFontSettings=gasp -Dawt.useSystemAAFontSettings=gasp 
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/usr/share/groovy/lib/groovy-2.5.13.jar) to method java.lang.Object.finalize()
WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Error: Diff mode may not use '--sevntuVersion' or '--checkstyleVersion' arguments!
Caught: java.lang.IllegalArgumentException: Error: invalid command line arguments!
...

➜  checkstyle-tester git:(issue-529) ✗ groovy diff.groovy -l projects-to-test-on.properties -r ../../checkstyle -p my-issue --sevntuVersion "8.41.1"

Picked up _JAVA_OPTIONS: -Dawt.useSystemAAFontSettings=gasp -Dawt.useSystemAAFontSettings=gasp 
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/usr/share/groovy/lib/groovy-2.5.13.jar) to method java.lang.Object.finalize()
WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Error: Diff mode may not use '--sevntuVersion' or '--checkstyleVersion' arguments!
Caught: java.lang.IllegalArgumentException: Error: invalid command line arguments!
...

```

File: checkstyle-tester/diff.groovy
Comment (romani) [inline checkstyle-tester/diff.groovy:68]: I think it is better to reuse extraMvnOptions.
as introduction of arguments that works in one mode only is hard to describe in doc and use.
By extraMvnOptions we are flexible.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: checkstyle-tester/diff.groovy
Comment (romani) [reply checkstyle-tester/diff.groovy:53]: sevntu usage should NOT change design on diff.groovy
it is just extension of maven config, if user add something in config he need to make sure that versions are overidable by maven variables.

File: checkstyle-tester/README.md
Comment (nmancus1) [reply checkstyle-tester/README.md]: These options have been removed now.

File: checkstyle-tester/README.md
Comment (nmancus1) [reply checkstyle-tester/README.md]: This option has been removed now.

File: checkstyle-tester/diff.groovy
Comment (nmancus1) [reply checkstyle-tester/diff.groovy]: Reverted this.

File: checkstyle-tester/diff.groovy
Comment (nmancus1) [reply checkstyle-tester/diff.groovy]: Reverted.

File: checkstyle-tester/diff.groovy
Comment (nmancus1) [reply checkstyle-tester/diff.groovy]: Done.

File: checkstyle-tester/diff.groovy
Comment (nmancus1) [reply checkstyle-tester/diff.groovy]: Done.

File: checkstyle-tester/diff.groovy
Comment (nmancus1) [reply checkstyle-tester/diff.groovy:66]: Done.

**Comment by nmancus1** [comment]
All items have been addressed up to this point, we need to make sure that all CI in other projects is passing. I will share results here.

**Comment by nmancus1** [comment]
I had to remove the diff report CI job; every time that we change a check we would have to update this job to a new commit, or it will fail. I am leaving the other job, since we are only generating a report for a single commit so that it will be consistent.

**Review by romani — APPROVED** [review]
This update is huge leap forward, there a lot items to make better, but lets not spend bunch of time on this polishing now, we can improve later on in other PRs.

**Comment by romani** [comment]
I got offline approval from @rnveach  to move forward

**Pull request merged**
